### PR TITLE
Changes for Azure Test Plan Task

### PR DESCRIPTION
**Pull request by adityashahms**
Making changes to create test run with test plan information and populating manual test point information for associated test case results parsed from test result files

Testing Screenshots:

- Task : 
image

- Test Plan Run Tab:
image

- Test Tab 
image

- Test Plan Execute Tab
image

- Test Point Pane
image

- Test Point Execution History
image

### CODE REVIEW

**Comment by adityashahms** [comment]
Feature Flag Implementation has introduced duplicate code. Once the feature flag is rollout, the function signature would be changed and code will be cleaned up.

**Comment by dogirala** [comment]
lgtm

**Comment by triptijain2112** [comment]
Can we check what is the test run name created in case of vstest?

File: src/Agent.Worker/TestResults/Utils/TestResultsConstants.cs [hunk 1]
```diff
@@ -24,5 +24,7 @@ public class TestResultsConstants
         public static readonly string EnableXUnitHeirarchicalParsing = "TestManagement.PublishTestResultsTask.EnableXUnitHeirarchicalParsing";
 
         public static readonly string JUnitTestCaseAttachmentsEnabled = "TestManagement.Server.JUnitTestCaseAttachmentsEnabled";
+
+        public static readonly string EnableAzureTestPlanTaskFeatureFlag = "TestManagement.Server.EnableAzureTestPlanTaskFlow";
```
Comment (triptijain2112) [inline src/Agent.Worker/TestResults/Utils/TestResultsConstants.cs:28]: Where is this FF added at ADO end?

**Review by triptijain2112 — APPROVED** [review]

File: src/Agent.Worker/TestResults/ResultsCommandExtension.cs [hunk 2]
```diff
@@ -260,23 +309,27 @@ private PublishOptions GetPublishOptions()
...
+        private async Task PublishTestRunDataAsync(string teamProject, TestRunContext testRunContext)
         {
             bool isTestRunOutcomeFailed = false;
 
-            var connection = WorkerUtilities.GetVssConnection(_executionContext);
-            var featureFlagService = _executionContext.GetHostContext().GetService<IFeatureFlagService>();
-            featureFlagService.InitializeFeatureService(_executionContext, connection);
-            var publishTestResultsLibFeatureState = featureFlagService.GetFeatureFlagState(TestResultsConstants.UsePublishTestResultsLibFeatureFlag, TestResultsConstants.TFSServiceInstanceGuid);
             _telemetryProperties.Add("UsePublishTestResultsLib", publishTestResultsLibFeatureState);
+            var connection = WorkerUtilities.GetVssConnection(_executionContext);
```
Comment (DergachevE) [inline src/Agent.Worker/TestResults/ResultsCommandExtension.cs:317]: Using? VssConnection implements IDisposable

File: src/Agent.Worker/TestResults/TestDataPublisher.cs [hunk 3]
```diff
@@ -117,6 +120,108 @@ public async Task<bool> PublishAsync(TestRunContext runContext, List<string> tes
...
+                                testResultByFQN[testResult.AutomatedTestName] = new List<TestCaseResult>();
+                            }
+                            // Add the object to the dictionary using its Id as the key
+                            testResultByFQN[testResult.AutomatedTestName].Add(testResult);
+                        }
+
+                        int testRunDataIterator = 0;
+                        int testResultDataIterator = 0;
+
+                        for (testRunDataIterator = 0; testRunDataIterator < testRunData.Count; testRunDataIterator++)
```
Comment (DergachevE) [inline src/Agent.Worker/TestResults/TestDataPublisher.cs:154]: How big are these sets of test run data and test results?

Comment (adityashahms) [reply src/Agent.Worker/TestResults/TestDataPublisher.cs:154]: We have threshold for test run data as 100, if it more than 100, we merge them and 1 test run data is formed. For Test results, it is on average around 100 and the worst seen so far is 1000.

File: src/Agent.Worker/TestResults/ResultsCommandExtension.cs (hunk 2 above)
Comment (adityashahms) [reply src/Agent.Worker/TestResults/ResultsCommandExtension.cs:317]: Follow up PR for implementing using statement for the connection to dispose the connection post completion of the task.

File: src/Agent.Worker/TestResults/ResultsCommandExtension.cs [hunk 4]
```diff
@@ -46,6 +47,11 @@ public sealed class PublishTestResultsCommand : IWorkerCommand
         private string _configuration;
         private string _runTitle;
         private bool _publishRunLevelAttachments;
+        private TestCaseResult[] _testCaseResults;
+        private string _testPlanId;
+        private bool enableAzureTestPlanFeatureState;
```
[the remark this answers is not in the document]
Comment (adityashahms) [reply src/Agent.Worker/TestResults/ResultsCommandExtension.cs:52]: Updated in this follow up PR : https://github.com/microsoft/azure-pipelines-agent/pull/4708

**Pull request merged**